src/Api/PubnubApi/EndPoint/Files/PublishFileMessage.cs

```diff
@@ -1,282 +1,281 @@
+				PNStatus errStatus = new PNStatus { Error = true, ErrorData = new PNErrorData("Missing Channel", new ArgumentException("Missing Channel")) };
+				returnValue.Status = errStatus;
+				return returnValue;
+			}
+
+			var requestParameter = CreateRequestParameter();
+			RequestState<PNPublishFileMessageResult> requestState = new RequestState<PNPublishFileMessageResult>();
+			requestState.ResponseType = PNOperationType.PNPublishFileMessageOperation;
+			requestState.Reconnect = false;
+			requestState.EndPointOperation = this;
+			Tuple<string, PNStatus> JsonAndStatusTuple;
+			var transportRequest = PubnubInstance.transportMiddleware.PreapareTransportRequest(requestParameter: requestParameter, operationType: PNOperationType.PNPublishFileMessageOperation);
+			var transportResponse = await PubnubInstance.transportMiddleware.Send(transportRequest);
+			if (transportResponse.Error == null) {
+				string responseString = Encoding.UTF8.GetString(transportResponse.Content);
+				PNStatus errorStatus = GetStatusIfError(requestState, responseString);
+				if (errorStatus == null) {
+					PNStatus status = new StatusBuilder(config, jsonLibrary).CreateStatusResponse(requestState.ResponseType, PNStatusCategory.PNAcknowledgmentCategory, requestState, (int)HttpStatusCode.OK, null);
+					JsonAndStatusTuple = new Tuple<string, PNStatus>(responseString, status);
+				} else {
+					JsonAndStatusTuple = new Tuple<string, PNStatus>(string.Empty, errorStatus);
+				}
+				returnValue.Status = JsonAndStatusTuple.Item2;
+			string json = JsonAndStatusTuple.Item1;
+			if (!string.IsNullOrEmpty(json)) {
+				List<object> result = ProcessJsonResponse(requestState, json);
+
+				if (result != null && result.Count >= 3) {
+					int publishStatus;
+					var _ = int.TryParse(result[0].ToString(), out publishStatus);
+					if (publishStatus == 1) {
+						List<object> resultList = ProcessJsonResponse(requestState, json);
+						if (resultList != null && resultList.Count > 0) {
+							ResponseBuilder responseBuilder = new ResponseBuilder(config, jsonLibrary, pubnubLog);
+							PNPublishFileMessageResult responseResult = responseBuilder.JsonToObject<PNPublishFileMessageResult>(resultList, true);
+							if (responseResult != null) {
+								returnValue.Result = responseResult;
+							}
+						}
+					}
+				}
+			}
+			} else {
+				int statusCode = PNStatusCodeHelper.GetHttpStatusCode(transportResponse.Error.Message);
+				PNStatusCategory category = PNStatusCategoryHelper.GetPNStatusCategory(statusCode, transportResponse.Error.Message);
+				PNStatus status = new StatusBuilder(config, jsonLibrary).CreateStatusResponse(PNOperationType.PNPublishFileMessageOperation, category, requestState, statusCode, new PNException(transportResponse.Error.Message, transportResponse.Error));
+				returnValue.Status = status;
+			}
+
+			return returnValue;
+		}
+
+		private RequestParameter CreateRequestParameter()
```

Comment:
I'm seeing CreateRequestParameter in a lot of these updated classes, is there some common functionality here that can be put in a base method PubnubCoreBase or is it completely different each time and not worth it?

Reply:
With this refactoring,   Each endpoint file contains code to generate the request url components.

So I just set a convention calling it a method name CreateRequestParameter (subject to change if we agree on some other name). Which is just an encapsulation of generating http request path segments and query param.

Since all endpoints have their own feature specific implementation It has to be endpoint file specific implementation. Also, It's private method for internal use. That's why It's not abstracted.    You can think of as a helper kind of method.